### PR TITLE
refactor: collections and attributes representation in `DataSet`s

### PR DESCRIPTION
**Pull request by kswenson**
Substantial refactor of the collection-related code and the way attributes are represented internally. This is part one of a larger refactor where the second part will tackle the way cases are represented.
- eliminate the `ungrouped` `CollectionPropsModel`
- eliminate the `CollectionPropsModel` type
- `DataSet`s always have at least one full-fledged `CollectionModel`
- `DataSet`'s `childCollection` property can be used where `ungrouped` was used previously
- `DataSet`'s `parentCollections` property can be used to access non-child collections
- as before, each collection's `attributes` property is an array of references
- the `DataSet`'s `attributes` property is now a computed property built from the individual collections
- eliminate `LEGACY_ATTRIBUTES_ARRAY_ANY` type in favor of a new `createDataSet` utility function

### CODE REVIEW

File: v3/src/models/data/data-set.ts [hunk 1]
```diff
@@ -483,83 +464,105 @@ export const DataSet = V2Model.named("DataSet").props({
...
+      }
+      dstCollection.addAttribute(attribute, options)
+      // update grouping
+      self.invalidateCollectionGroups()
+    }
+    return { removedCollectionId }
+  }
+}))
+.actions(self => ({
+  // if beforeCollectionId is not specified, new collection is last (child-most)
```
Comment (tealefristoe) [inline v3/src/models/data/data-set.ts:543]: Is this comment accurate? Looking at `addCollection`, it looks like the default is to add left of the child-most collection. For what it's worth, I think adding either a child-most or parent-most collection would be better than left of the child-most collection.

**Review by tealefristoe — APPROVED** [review]
Nice! I made a few minor comments but overall this looks great and I can't wait to work with the changes.

File: v3/src/components/map/components/map-point-layer.tsx [hunk 2]
```diff
@@ -145,7 +145,9 @@ export const MapPointLayer = observer(function MapPointLayer({mapLayerModel, onS
   const refreshConnectingLines = useCallback(() => {
     if (!showConnectingLines && !connectingLinesActivatedRef.current) return
     const connectingLines = connectingLinesForCases()
-    const parentAttr = dataset?.collections[0]?.attributes[0]
+    // TODO: is this the right rule for parent grouping attribute? Seems like it
+    // should depend on the collection(s) in which the plotted attributes reside.
+    const parentAttr = (dataset?.collections.length ?? 0) > 1 ? dataset?.collections[0]?.attributes[0] : undefined
```
[the remark this answers is not in the document]
Comment (kswenson) [reply v3/src/components/map/components/map-point-layer.tsx]: Added a `firstVisibleParentAttribute()` function to `data-set-utils.ts`.

File: v3/src/models/data/collection.ts [hunk 3]
```diff
@@ -13,10 +13,37 @@ export const CollectionLabels = types.model("CollectionLabels", {
...
+  getAttribute(attrId: string) {
+    return self.attributes.find(attribute => attribute?.id === attrId)
+  },
+  getAttributeIndex(attrId: string) {
+    return self.attributes.findIndex(attribute => attribute?.id === attrId)
+  },
+  getAttributeByName(name: string) {
+    return self.attributes.find(attribute => attribute?.name === name)
+  },
+  get caseIdMap() {
```
[the remark this answers is not in the document]
Comment (kswenson) [reply v3/src/models/data/collection.ts]: Done

File: v3/src/models/data/data-set.ts [hunk 4]
```diff
@@ -483,83 +464,105 @@ export const DataSet = V2Model.named("DataSet").props({
...
+        if (beforeIndex < 0 && self.collections.length > 0) {
+          beforeIndex = self.collections.length - 1
+        }
         if (beforeIndex >= 0) {
           self.collections.splice(beforeIndex, 0, collection)
         }
         else {
           self.collections.push(collection)
         }
+        const newCollection = self.collections[beforeIndex >= 0 ? beforeIndex : self.collections.length - 1]
```
[the remark this answers is not in the document]
Comment (kswenson) [reply v3/src/models/data/data-set.ts]: Fixed

File: v3/src/models/data/data-set.ts (hunk 1 above)
Comment (kswenson) [reply v3/src/models/data/data-set.ts]: I experimented with changing the default but it was going to involve changing a bunch of tests, etc. so I decided to leave it alone and just fix the comment. For what it's worth, a logic for the current behavior is that if you think of the attributes of the child-most collection as the ungrouped attributes, adding a new collection is like pulling an attribute out of the ungrouped attributes and the most logical place to put that new collection is immediately adjacent to the child-most collection.

**Pull request merged**